ports/servoshell/headless_window.rs

```diff
@@ -74,14 +67,21 @@ impl WindowPortsMethods for Window {
         unsafe { winit::window::WindowId::dummy() }
     }
 
+    fn hidpi_factor(&self) -> Scale<f32, DeviceIndependentPixel, DevicePixel> {
+        match self.device_pixel_ratio_override {
+            Some(override_value) => Scale::new(override_value),
+            _ => Scale::new(1.0),
+        }
+    }
+
```

Comment:
It seems like the implemetnation of this method is very similar for headed and headless windows. Could this be implemented as a [default method on the trait](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations) that uses two non-default methods: `device_pixel_ratio_override()` and `device_pixel_ratio()` (which would return 1.0) for headless windows)?

Reply:
headed Window also has a case for `opts::get().output_file` that headless Window lacks, so i think this would be a bit of a leaky/premature abstraction. what do you think?

Reply:
I think it's fine to move the `opts::get().output_file` case to the default method. It doesn't seem like this behavior should differ between the two types of windows. My suspicion is that the only reason it's only for headed windows is that headless windows default to 1.0 anyway. Moving it to the default method will make it much clearer what happens when you specify both options though.